### ISSUE TITLE
Optimization failure due to: local_reduce_broadcastable

### ISSUE THREAD

**Issue by ppalasek** [Theano/Theano#4300]
Hello,

I am getting the following error when compiling my function:

```
<<!! BUG IN FGRAPH.REPLACE OR A LISTENER !!>> <type 'exceptions.TypeError'> ('The type of the replacement must be compatible with the type of the orig
inal Variable.', Sum{axis=[0, 2, 3], acc_dtype=float64}.0, Sum{axis=[0], acc_dtype=int64}.0, TensorType(float64, vector), TensorType(int64, vector), '
local_reduce_broadcastable') local_reduce_broadcastable
ERROR (theano.gof.opt): Optimization failure due to: local_reduce_broadcastable
ERROR (theano.gof.opt): node: Sum{axis=[0, 2, 3], acc_dtype=float64}(DimShuffle{0,3,1,2}.0)
ERROR (theano.gof.opt): TRACEBACK:
ERROR (theano.gof.opt): Traceback (most recent call last):
  File "/homes/pp305/work/theano_code/theano_new/lib/python2.7/site-packages/theano/gof/opt.py", line 1807, in process_node
    fgraph.replace_all_validate(repl_pairs, reason=lopt)
  File "/homes/pp305/work/theano_code/theano_new/lib/python2.7/site-packages/theano/gof/toolbox.py", line 309, in replace_all_validate
    fgraph.replace(r, new_r, reason=reason, verbose=False)
  File "/homes/pp305/work/theano_code/theano_new/lib/python2.7/site-packages/theano/gof/fg.py", line 561, in replace
    str(reason))
TypeError: ('The type of the replacement must be compatible with the type of the original Variable.', Sum{axis=[0, 2, 3], acc_dtype=float64}.0, Sum{ax
is=[0], acc_dtype=int64}.0, TensorType(float64, vector), TensorType(int64, vector), 'local_reduce_broadcastable')
```

Theano version I'm using is:
0.8.0.dev-fba8851ca5b2d6aa70f89c7b6e1343e55a7762c3

I was running the code on CPU.

Any suggestions on how to fix this?

It's quite a messy code I'm running at the moment so I don't know the exact source of the error, but I can try to narrow it down in the next few days.

Best,
Petar

**Comment by nouiz** [comment]
You can ignore this. Theano will ignore that trouble some case.

Can you give a function_dump() so that we investigate this?

http://www.deeplearning.net/software/theano/library/compile/function.html#theano.compile.function.function_dump

thanks

**Comment by ppalasek** [comment]
Hi nouiz,

the function dump is attached: [one_mini_batch_step_dump.pkl.zip](https://github.com/Theano/Theano/files/198145/one_mini_batch_step_dump.pkl.zip)

When I changed the type of one variable from itensor3() to ftensor3() (so that all the computation is done with float32) the error didn't appear. But in this case I need to cast some values to int32 later when I use them in e.g. indexing something.

I guess the error might be related to T.tensordot, but that's just a guess.

Let me know if you need any other info.

Best,
Petar

**Comment by nouiz** [comment]
It seem we dropped this issue.

I looked into it. Currently Theano give a different error message that is better then this one.

The problem is that in your givens parameter, the replacement you do don't have the same broadcast pattern.

(Pdb) p list(d["givens"].iteritems())[0][0].type
TensorType(int32, 3D)
(Pdb) p list(d["givens"].iteritems())[0][1].type
TensorType(int32, (False, True, True))


So I'll close this issue.
